### ISSUE TITLE
Typecheck breaks for response schema not containing 200 status code

### ISSUE THREAD

**Issue by aburii** [elysiajs/elysia#785]
### What version of Elysia is running?

latest

### What platform is your computer?

Darwin 23.5.0 arm64 arm

### What steps can reproduce the bug?

```typescript
import { Elysia, t } from "elysia";

const app = new Elysia()
    .get("/hello", () => "Hello Elysia", {
        response: {
            200: t.String()
        }
    })
    .get('/wrong', () => {
        return { 201: 1 } // Should be marked as error  as below
    }, {
        response: {
            201: t.String()
        }
    })
    .get('/wrong-200', () => {
        return { 200: 1 } // Raises the error
    }, {
        response: {
            200: t.String()
        }
    })
    .listen(3000);

console.log(
  `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
);
```

Here is the screenshot of my IDE (WebStorm 2024.2)

<img width="698" alt="image" src="https://github.com/user-attachments/assets/09477bb6-af74-4137-8db7-1eb19fdc7eeb">

### What is the expected behavior?

The expected behavior is that all 20(X) (201, 204 and so on) typecheck work.

In this case
```typescript
    .get('/wrong', () => {
        return { 201: 1 }
    }, {
        response: {
            201: t.String()
        }
    })
```

Should raise the same error as for 200 status code like: 

<img width="375" alt="image" src="https://github.com/user-attachments/assets/7c27e5fa-cdbf-4b23-bf5b-aa85b096ff45">

Details: 

<img width="933" alt="image" src="https://github.com/user-attachments/assets/01864146-e426-4204-a68d-0626c645b043">

### What do you see instead?

201 response schema does not trigger typecheck error raise.

### Additional information

I am 100% this error does NOT come from my LSP nor my IDE since I have cleaned up the cache and restarted the services multiples times.

It seems that the pasts fixes have been removed (e.g. PR #596 ) 

### Have you try removing the `node_modules` and `bun.lockb` and try again yet?

Yes

**Comment by bogeychan** [comment]
The type behaviour is correct.

`201` is the status code, no response body value.

```js
response: {
    201: t.String()
}
```

Based on your example, read the comments:

```ts
new Elysia()
  .get(
    "/hello",
    () => "Hello Elysia", // no error because you return string with 200 status code
    {
      response: {
        200: t.String(),
      },
    }
  )
  .get(
    "/wrong",
    () => {
      return { 201: 1 }; // no error because you implicit return 200 status code with json body of { 201: 1 }
    },
    {
      response: {
        201: t.String(),
      },
    }
  )
  .get(
    "/wrong-200",
    () => {
      return { 200: 1 }; // This raises error because 200 response schema (see below) overwrites the default. and returning here should be string, not object (json)
    },
    {
      response: {
        200: t.String(),
      },
    }
  );
```

**Comment by bogeychan** [comment]
All this follows the OpenAPI `responses` semantic: https://swagger.io/docs/specification/describing-responses/

**Comment by aburii** [comment]
@bogeychan same issue here on this code

```typescript
import { Elysia, t } from "elysia";

const app = new Elysia()
    .get("/hello", () => "Hello Elysia", {
        response: {
            200: t.String()
        }
    })
    .get('/wrong', ({ set }) => {
        set.status = 'Created'
        return 1 // Should be marked as error  as below
    }, {
        response: {
            201: t.String()
        }
    })
    .get('/wrong-200', () => {
        return 1 // Raises the error
    }, {
        response: {
            200: t.String()
        }
    })
    .listen(3000);

console.log(
  `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
);
```

The first examples I've shown are indeed not correct.

Here is the screenshot of the current error state in my IDE (WebStorm 2024.2)

<img width="592" alt="image" src="https://github.com/user-attachments/assets/b4c27fbf-6c3f-4779-8e9a-1b89dac5bd9b">

I have to mention this issue has already been fixed by april but it seems to happen again, please refer to #596.

**Comment by bogeychan** [comment]
@aburii thanks for sharing this finding.

---

@SaltyAom
- The `response` schema with different status codes, as shown above (https://github.com/elysiajs/elysia/issues/785#issuecomment-2297342653), doesn't work anymore.
- And returning `number` with `response` schema of `t.String` shouldn't give a type error, maybe?

**Comment by aburii** [comment]
> @aburii thanks for sharing this finding.
> 
> @SaltyAom
> 
> * The `response` schema with different status codes, as shown above ([Typecheck breaks for response schema not containing 200 status code #785 (comment)](https://github.com/elysiajs/elysia/issues/785#issuecomment-2297342653)), doesn't work anymore.
> * And returning `number` with `response` schema of `t.String` shouldn't give a type error, maybe?

And please ignore the firsts examples I provided, they are incorrect, only the lasts are. By trying to illustrate more on the first examples I've shown no relevant issue.

What I tried to share are in fact the last examples.

**Comment by SaltyAom** [comment]
> @bogeychan same issue here on this code
> 
> ```ts
> import { Elysia, t } from "elysia";
> 
> const app = new Elysia()
>     .get("/hello", () => "Hello Elysia", {
>         response: {
>             200: t.String()
>         }
>     })
>     .get('/wrong', ({ set }) => {
>         set.status = 'Created'
>         return 1 // Should be marked as error  as below
>     }, {
>         response: {
>             201: t.String()
>         }
>     })
>     .get('/wrong-200', () => {
>         return 1 // Raises the error
>     }, {
>         response: {
>             200: t.String()
>         }
>     })
>     .listen(3000);
> 
> console.log(
>   `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
> );
> ```
> 
> The first examples I've shown are indeed not correct.
> 
> Here is the screenshot of the current error state in my IDE (WebStorm 2024.2)
> 
> <img alt="image" width="592" src="https://private-user-images.githubusercontent.com/71981971/359228754-b4c27fbf-6c3f-4779-8e9a-1b89dac5bd9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NDc0MjYsIm5iZiI6MTcyNDc0NzEyNiwicGF0aCI6Ii83MTk4MTk3MS8zNTkyMjg3NTQtYjRjMjdmYmYtNmMzZi00Nzc5LThlOWEtMWI4OWRhYzViZDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA4MjUyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjNjAzYzg3ZGNiMzExMzZlYjNhN2VjN2QzNjA5OTkyZWUwYjNhM2Y3NTVhMzllYmNlMjQzNmE4MDgwY2U2N2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BI3068WQeOnfE3AzKmq53uWjpkYycVoZ3PY_1-557zM">
> I have to mention this issue has already been fixed by april but it seems to happen again, please refer to #596 and #490

The `/wrong` here is not possible due to limitation of TypeScript. As it's not possible to infer the type of `set.status = 201`, Elysia assume the endpoint returning 200 status code.

To make this possible, Elysia has a dedicated [`error` function](https://elysiajs.com/essential/handler.html#error) to validate type integrity.

<img width="741" alt="Screenshot 2567-08-27 at 15 28 31" src="https://github.com/user-attachments/assets/bddf065d-73c6-46a6-a5fa-ea86a5e94e9d">

So if we refactor the code to use `error` function as the following:
```typescript
new Elysia()
    .get('/wrong', ({ error }) => {
        return error(201, 1)
    }, {
        response: {
            201: t.String()
        }
    })
    .listen(3000)
```

We can see that the IDE now show the error.
<img width="996" alt="Screenshot 2567-08-27 at 15 28 54" src="https://github.com/user-attachments/assets/11bc767a-980b-454c-8331-e0365a82fee6">

So in other words, I recommended to use a dedicated `error` function instead of using `set.status` to ensure type integrity as using `set.status` isn't possible to infer return type due to TypeScript limitation.

**Comment by aburii** [comment]
@SaltyAom Feels a bit weird to call a function 'error' to send back a success status, is there any function planned on next versions of Elysia to feel _more accurate_ ?

**Comment by tkesgar** [comment]
I encountered this yesterday and ended up aliasing `error` using resolve hook (since I still need `error` to send, well, errors):

```ts
const app = new Elysia()
  .resolve(({ error }) => ({ send: error }))
  .post(({ error, send }) => {
    if (postAlreadyExists) {
      return error(409, { message: 'Post already exists' })
    } else {
      return send(201, { message: 'Post created successfully' })
    }
  })
```

**Comment by SaltyAom** [comment]
Closing as complete with the following

> > [@bogeychan](https://github.com/bogeychan) same issue here on this code
> > import { Elysia, t } from "elysia";
> > 
> > const app = new Elysia()
> >     .get("/hello", () => "Hello Elysia", {
> >         response: {
> >             200: t.String()
> >         }
> >     })
> >     .get('/wrong', ({ set }) => {
> >         set.status = 'Created'
> >         return 1 // Should be marked as error  as below
> >     }, {
> >         response: {
> >             201: t.String()
> >         }
> >     })
> >     .get('/wrong-200', () => {
> >         return 1 // Raises the error
> >     }, {
> >         response: {
> >             200: t.String()
> >         }
> >     })
> >     .listen(3000);
> > 
> > console.log(
> >   `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
> > );
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The first examples I've shown are indeed not correct.
> > Here is the screenshot of the current error state in my IDE (WebStorm 2024.2)
> > <img alt="image" width="592" src="https://private-user-images.githubusercontent.com/71981971/359228754-b4c27fbf-6c3f-4779-8e9a-1b89dac5bd9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NDc0MjYsIm5iZiI6MTcyNDc0NzEyNiwicGF0aCI6Ii83MTk4MTk3MS8zNTkyMjg3NTQtYjRjMjdmYmYtNmMzZi00Nzc5LThlOWEtMWI4OWRhYzViZDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA4MjUyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjNjAzYzg3ZGNiMzExMzZlYjNhN2VjN2QzNjA5OTkyZWUwYjNhM2Y3NTVhMzllYmNlMjQzNmE4MDgwY2U2N2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BI3068WQeOnfE3AzKmq53uWjpkYycVoZ3PY_1-557zM">
> > I have to mention this issue has already been fixed by april but it seems to happen again, please refer to [#596](https://github.com/elysiajs/elysia/pull/596) and [#490](https://github.com/elysiajs/elysia/issues/490)
> 
> The `/wrong` here is not possible due to limitation of TypeScript. As it's not possible to infer the type of `set.status = 201`, Elysia assume the endpoint returning 200 status code.
> 
> To make this possible, Elysia has a dedicated [`error` function](https://elysiajs.com/essential/handler.html#error) to validate type integrity.
> 
> <img alt="Screenshot 2567-08-27 at 15 28 31" width="741" src="https://private-user-images.githubusercontent.com/35027979/361719080-bddf065d-73c6-46a6-a5fa-ea86a5e94e9d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQyNjQyODIsIm5iZiI6MTc1NDI2Mzk4MiwicGF0aCI6Ii8zNTAyNzk3OS8zNjE3MTkwODAtYmRkZjA2NWQtNzNjNi00NmE2LWE1ZmEtZWE4NmE1ZTk0ZTlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAzVDIzMzMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkODQxOTM3YjUwMmQzZTMyOTk1ZWM1ODU0NjE2ZWYwMDkxZDA1MTQwYzFlYTEzNTZhYTc3NzM0MjQzNDQxYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JISXMMCxe_QzOHNEcIYRzAZydzENuDOgW6V1K6o-XGI">
> So if we refactor the code to use `error` function as the following:
> 
> new Elysia()
>     .get('/wrong', ({ error }) => {
>         return error(201, 1)
>     }, {
>         response: {
>             201: t.String()
>         }
>     })
>     .listen(3000)
> We can see that the IDE now show the error. <img alt="Screenshot 2567-08-27 at 15 28 54" width="996" src="https://private-user-images.githubusercontent.com/35027979/361719409-11bc767a-980b-454c-8331-e0365a82fee6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQyNjQyODIsIm5iZiI6MTc1NDI2Mzk4MiwicGF0aCI6Ii8zNTAyNzk3OS8zNjE3MTk0MDktMTFiYzc2N2EtOTgwYi00NTRjLTgzMzEtZTAzNjVhODJmZWU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODAzVDIzMzMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyNzRlYzk0ZTAzOTMyNjhmZmNkZTVjMTFiOWNjY2ZmMjJlZWU0ZmM5MjQ0NjhmMDcxNzc2MGU5OTRkOGRmZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qNthcxx722dLX_BGYwKPsT_2lFth7aY7vA0K_tjK51g">
> 
> So in other words, I recommended to use a dedicated `error` function instead of using `set.status` to ensure type integrity as using `set.status` isn't possible to infer return type due to TypeScript limitation.

You can now use `status` which is a nwer alias of `error` to make it doesn't feels weird
